Step group template (#625)
harness/developer-hub: docs/platform/13_Templates/create-a-stepgroup-template.md

---
title: Create a Step Group Template
description: Streamline and standardize your deployment processes with step group templates.
sidebar_position: 6

---

> A step group in Harness is a collection of deployment steps that are executed together as a single unit. This allows you to organize and manage complex deployment processes by dividing them into smaller, manageable steps. A step group can contain any type of deployment step, including script steps, artifact deployment steps, and infrastructure deployment steps. You can also control the order in which the steps in a step group are executed. In addition, you can specify conditions for skipping certain steps based on the outcome of previous steps. Overall, step groups are a useful tool for streamlining and organizing your deployment processes in Harness.

Reviewer: any type of deployment step -> we should say any type of Harness Pipeline Step. deployment is specific to CD. CI has different terminology

Author: Done

> By creating a step group template, you can standardize and simplify your deployment processes, and reduce the time and effort required to create new deployments. You can parameterize certain aspects of the template to allow for customization when it is used in a deployment. Additionally, you can apply conditions to the steps in a step group template to make it more flexible and reusable in different deployment scenarios. 

Reviewer: to create new deployments -> to create new pipelines and templates

Author: Done

> You can share your step group template at the project, organization, and account levels.

Reviewer: You can share your -> You can create your

Author: Done

## Create a step group template

> First, we'll create a project-level template in the **Deployments** module. You can do this in any project.

Reviewer: project-level -> project level

Author: Rephrased to "Let us start by creating a template in the project scope within the **Deployments** module. "

6. In **Version Label**, enter the version of the stage. You can enter v1.
7. Select the scope at which you want to save this template in **Save To**.
8. To set up your template, select one of the following: 
   - **Inline**: Save the template in Harness.
   - **Remote**: Save the template in a Git repository.
  
  
### Create a remote template

  1. In **Git Connector**, select or create a Git Connector to the repo for your project.​ For steps, see [Code Repo Connectors](../7_Connectors/connect-to-code-repo.md).
  2. In **Repository**, select your repository.​ If your repository isn't listed, enter its name since only a select few repositories are filled here.​Create the repository in Git before entering it in Select Repository. ​Harness does not create the repository for you.​
  3. In **Git Branch**, select your branch.​ If your branch isn't listed, enter its name since only a select few branches are filled here.Create the branch in your repository before entering it in Git Branch. ​Harness does not create the branch for you​​
  4. ​Harness auto-populates the **YAML Path**.​ You can change this path and the file name.
  5. Click **Start**.

## Select the stage type

> 1. Select **Deploy**. The deploy stage type is a CD Stage that enables you to deploy any service to your target environment.

Reviewer: can we mention that this deploy type decided where this step group template can be used ?
if type if deploy, it can used be used in "deploy" stage or stage templates
if type if build, it can used be used in "build" stage or stage templates

Author: "The stage type you choose determines where you can reference the step group template in the pipeline. If you select Deploy as the stage type, you can reference the template in any pipeline's Deploy stage. If you select Build as the stage type, you can reference the template in any pipeline's Build stage."

---
title: Create a Step Group Template
description: Streamline and standardize your deployment processes with step group templates.
sidebar_position: 6

---

> A step group in Harness is a collection of deployment steps that are executed together as a single unit. This allows you to organize and manage complex deployment processes by dividing them into smaller, manageable steps. A step group can contain any type of pipeline step, including script steps, artifact deployment steps, and infrastructure deployment steps. You can also control the order in which the steps in a step group are executed. In addition, you can specify conditions for skipping certain steps based on the outcome of previous steps. Overall, step groups are a useful tool for streamlining and organizing your deployment processes in Harness.

Author: These are two words; the standard way to include them is without hyphens.

By creating a step group template, you can standardize and simplify your deployment processes, and reduce the time and effort required to create new pipelines and templates. You can parameterize certain aspects of the template to allow for customization when it is used in a deployment. Additionally, you can apply conditions to the steps in a step group template to make it more flexible and reusable in different deployment scenarios. 

You can create your step group template at the project, organization, and account levels.

## Create a step group template

Let us start by creating a template in the project scope within the **Deployments** module.  You can do this in any project.

1. Navigate to the **Deployments** module and in **Projects** select the desired project.
2. Next select **Templates** under **Project Setup**.
3. Click **New Template**.
4. Select **Step Group** to create a stage template.
   
   ![](./static/step-group-template-option.png)

   The **Create New Step Group Template** settings appear.

5. In **Name**, enter a name for the template.
6. In **Version Label**, enter the version of the stage. You can enter v1.

> 7. Select the scope at which you want to save this template in **Save To**.

Author: Done

8. To set up your template, select one of the following: 
   - **Inline**: Save the template in Harness.
   - **Remote**: Save the template in a Git repository.
  
  
### Create a remote template

  1. In **Git Connector**, select or create a Git Connector to the repo for your project.​ For steps, see [Code Repo Connectors](../7_Connectors/connect-to-code-repo.md).

>   2. In **Repository**, select your repository.​ If your repository isn't listed, enter its name since only a select few repositories are filled here.​Create the repository in Git before entering it in Select Repository. ​Harness does not create the repository for you.​

Author: It is "Repository".

  3. In **Git Branch**, select your branch.​ If your branch isn't listed, enter its name since only a select few branches are filled here.Create the branch in your repository before entering it in Git Branch. ​Harness does not create the branch for you​​
  4. ​Harness auto-populates the **YAML Path**.​ You can change this path and the file name.
  5. Click **Start**.

## Select the stage type

1. Select **Deploy**. The deploy stage type is a CD Stage that enables you to deploy any service to your target environment.

   ![](./static/stage-type.png)

   You can also select other types of stages. This topic uses the **Deploy** stage type.
   

>    The stage type you choose determines where you can reference the step group template in the pipeline. If you select **Deploy** as the stage type, you can reference the template in any pipeline's deploy stage. If you select **Build** as the stage type, you can reference the template in any pipeline's build stage.

Author: Rephrased a bit differently.

2. In your stage, click **Add Step** and then click **Add Step**.

   Select a step type from the step library.

   Let us add a **Canary Deployment** step.

   Enter a name for this step and then click **Apply Changes**.

3. Next, let us add an approval step.

   Click **Add Step** and then click **Add Step**.

   In **Approval**, select **Harness Approval**.

   Enter a name for this step. Select the user groups and then click **Apply Changes**.

>    You can add more steps to form a step group.

Author: The intention here is to tell the users that they can add more steps if they want. Step group in itself is a group of multiple steps.

                approvers:
                userGroups:
                    - account._account_all_users
                minimumCount: 1
                disallowPipelineExecutor: false
                approverInputs: []
            timeout: 1d
        - step:
            type: K8sRollingDeploy
            name: Rolling Deployment
            identifier: Rolling_Deployment
            spec:
                skipDryRun: false
                pruningEnabled: false
            timeout: 10m
        stageType: Deployment
        when:
        stageStatus: Success
        condition: <+input>
    identifier: Validate_Kubernetes_Rollout
    versionLabel: "1.0"
    ```

4. Click **Save** to save the step group template you just created.

## Add details for conditional execution

A step group can have its own **Conditional Execution** settings separate from the Conditional Execution settings for the Stage. The Conditional Execution settings of the step group apply to all of its steps.

See Step Skip Condition Settings.

The Conditional Execution settings of any step in a step group overrides the conditional execution settings of the step group.

> If you do not use step group Conditional Execution settings, then the stage's conditional execution settings are used.

Reviewer 2: ```suggestion
If you do not use step group Conditional Execution settings, then the stage's conditional execution settings are applied.
```

## Add a failure strategy

A step group can have its own **Failure Strategy** separate from the failure strategy for the stage.

The failure strategy can execute the rollback steps for the step group.

> For more information, see [Step Failure Strategy Settings](../8_Pipelines/w_pipeline-steps-reference/step-failure-strategy-settings.md)

Reviewer 2: ```suggestion
For more information, go to [Step Failure Strategy Settings](../8_Pipelines/w_pipeline-steps-reference/step-failure-strategy-settings.md)
```

Author: This is a cross-reference from another topic. "See" would be appropriate here.

The Failure Strategy of any step in a step group overrides the Failure Strategy of the step group.

> If you do not use a step group Failure Strategy, then the Stage's Failure Strategy is used.

Reviewer 2: ```suggestion
If you do not use a step group Failure Strategy, then the Stage's Failure Strategy is applied.
```

## Add looping strategy

> Select one of the following **Looping Strategy** to run your step multiple times with different inputs: 

Reviewer 2: ```suggestion
Select one of the following **Looping Strategy** options to run a step multiple times with different inputs: 
```

- Matrix
- Repeat
- Parallelism

> For more information, see [Looping Strategies Overview](../8_Pipelines/looping-strategies-matrix-repeat-and-parallelism.md)

Reviewer 2: ```suggestion
For more information, go to [Looping Strategies Overview](../8_Pipelines/looping-strategies-matrix-repeat-and-parallelism.md)
```

> Click **Save**. Enter a comment for you changes and click **Save**.

Reviewer 2: ```suggestion
Click **Save**. Enter a comment for the changes and click **Save**.
```

> Click **Save**. Enter a comment for you changes and click **Save**.

Reviewer 2: This is confusing. Do I need to click Save and then a screen appears where I enter the comments, and then click Save once again?

Author: Yes.

## Add the step group template to a pipeline

> You can use the step group template in any pipeline in your project now that you have it.

Reviewer 2: You can use the step group template in any pipeline within your project.

Author: Done

To add a step group template to a pipeline: 

1. Go to the pipeline, and then select the stage where you want to add the template.
2. Click **Add Step**.
3. Select **Use template**.

> 4. Select your step group template from the list of templates.

Reviewer 2: ```suggestion
4. Select step group template from the list of templates.
```

Author: This is specific to the template that was created in the previous step.

5. Click **Use Template**.

> 6. Enter a name for your template and click **Apply Changes**.

Reviewer 2: ```suggestion
6. Enter a name for the template and click **Apply Changes**.
```

7. Click **Save**.

>    The template is now referenced in your pipeline.

Reviewer 2: Your pipeline now includes a reference to the template.

Author: Done
